Epinio journey: single dev workflow (#145)
epinio/docs: docs/tutorials/single-dev-workflow.md

---
sidebar_label: "Epinio Journey: Single Dev Workflow"
title: ""
---

## Introduction

While the [quickstart](./quickstart.md) is great to get started, as a Developer, you might want to see end-to-end solutions and how Epinio can help you.

> That's exactly the aim of the "Epinio Journeys", where you'll be to follow, according to your needs, different use-cases.

Reviewer: ```suggestion
That's exactly the aim of the "Epinio Journeys", where you'll be able to follow different use-cases, according to your needs.
```

> In this particular tutorial, we focus on the single dev workflow and go through a full example from Epinio installation to your first app deployment.

Reviewer 2: This might improve flow: 
```
In this particular tutorial, we focus on the workflow for a solo developer with an example from a bare Kubernetes deployment to your first app.
```

Reviewer 2: It might also be good to call out that this is for local development as well

## Prerequisites

Before you can use Epinio, you need to have a Kubernetes cluster running.

> As a single dev, you might be using a local Kubernetes cluster such as [Rancher Desktop](https://rancherdesktop.io/), [Minikube](https://minikube.sigs.k8s.io/docs/) or [k3d](https://k3d.io/), just to name a few.

Reviewer 2: s/single/solo/

Reviewer: ```suggestion
As a solo developer, you might be using a local Kubernetes cluster such as [Rancher Desktop](https://rancherdesktop.io/), [Minikube](https://minikube.sigs.k8s.io/docs/) or [k3d](https://k3d.io/), just to name a few.
```

:::note

This tutorial will not explain how to install a local Kubernetes cluster and assumes you know how to do it.

> However, depending on your choice, there's already some examples in **Epinio howtos** section which contains links to the respective clusters' installation docs.

Reviewer: ```suggestion
However, depending on your choice, there's already some examples in the **Epinio Howtos** section which contains links to the respective clusters' installation docs.
```

> However, depending on your choice, there's already some examples in the **Epinio Howtos** section which contains links to the respective clusters' installation docs.

Reviewer 2: Will the "Epinio Howtos" create a link to the section?

:::

In this tutorial, we'll use [Rancher Desktop](../howtos/install_epinio_on_rancher_desktop.md) as our local Kubernetes cluster. However you should be able to follow this tutorial with the local Kubernetes cluster of your choice.

If not already done, you can install the [latest version](https://github.com/rancher-sandbox/rancher-desktop/releases) of Rancher Desktop for your operating system.

### [Optional] Additional binaries

> Two additional binaries need to be installed in your system:

Reviewer 2: I think it might be good to be explicit that these additional binaries are only needed for the installation and customization. The actual dev flow doesn't need k8s tools

- [kubectl](https://kubernetes.io/docs/tasks/tools/#kubectl) for communicating with the Kubernetes cluster

- [helm](https://helm.sh/docs/intro/install/) for deploying the Epinio Helm Charts

Depending on the local Kubernetes cluster you installed (i.e. Rancher Desktop), these two binaries might be already installed.

## Installation

Once you have your local Kubernetes cluster installed and running, you can [install Epinio](../installation/install_epinio.md).

Here are the steps for Rancher Desktop:

```shell
kubectl apply -f https://github.com/cert-manager/cert-manager/releases/download/v1.7.1/cert-manager.yaml

> # Wait for cert-manager to stabilize

Reviewer 2: Should take < 30 seconds assuming a decent internet connection.

helm repo add epinio https://epinio.github.io/helm-charts

helm repo update

helm install epinio -n epinio --create-namespace epinio/epinio --set global.domain=127.0.0.1.sslip.io
```

### CLI

To interact with your Epinio installation, you've to download the [Epinio CLI binary](https://github.com/epinio/epinio/releases/latest). The binary is available for the three main Operating Systems (OS), so pick the one suited for your own OS.

### Login

The first task to perform after Epinio installation, is to [login](../references/commands/cli/epinio_login.md) with the binary you downloaded:

```shell

> epinio login -u admin 'https://epinio.127.0.0.1.sslip.io'

Reviewer 2: What's the password? 😃

Reviewer: The `NOTES.txt` printed at the end of deploying `epinio` with `helm` tells you.

# Trust the certificate by pressing 'y' and 'enter'
```

:::tip

If your local Kubernetes cluster restarts, you only need to login again with the command above. Epinio stays installed and the certificates are still valid.

:::

You can confirm that you're logged in by checking the Epinio settings:

```shell
epinio settings show
```

Alternatively, you can also open the Epinio URL, <https://epinio.127.0.0.1.sslip.io>, in your preferred browser and use the web UI.

## Deploy your application with Epinio

> Now that the "operational" tasks are done, it's time to concentrate on the most important task: use Epinio to deploy your application.

Reviewer 2: maybe s/deploy/develop/ ?

Reviewer: To me `develop` is the `act of writing the app`. And `deploy` is the act of running it on a cluster.
So I am with `deploy` here, not `develop`

> As [explained here](../explanations/detailed-push-process.md#7-stage), Epinio uses [Paketo buildpacks](https://paketo.io/) to create a container image for your application. This image is then used to create a container with your application, which will run on your local Kubernetes cluster.

Reviewer 2: I'd start with "Epinio uses Paketo..." and pull the "as explained here" to the end of the line as "more details about this can be found here"

Reviewer: I.e., likely
```suggestion
Epinio uses [Paketo buildpacks](https://paketo.io/) to create a container image for your application. This image is then used to create a container with your application, which will run on your local Kubernetes cluster. More about this can be found in the explanations of the [detailed push process](../explanations/detailed-push-process.md#7-stage).
```

Epinio will also create a new `ingress route`, which will allow you to easily access your application once it's deployed.

The whole process is handled by Epinio, which enables you to concentrate on your application rather than knowing how you'll be able to deploy it.

Let's see first how to deploy a simple application:

```shell
# Example code: https://github.com/epinio/example-12factor

# Move to the source code directory. Here is an example:
cd /github/example-12factor

# Deploy your application
epinio push -n mysimpleapp
```

At the end of the deployment output, you have the URL to be used for checking your application:

```shell
Deploying application ...
..............
🕞  Creating application resources
✔️  App is online.
Name: mysimpleapp
Namespace: workspace
Builder Image:
Routes:
1: https://mysimpleapp.127.0.0.1.sslip.io
```

### List the applications deployed

If you're working on many applications, it can be really useful to see when was the last time they were deployed and which URL you should use to check them.

> You can get the applications information with the following two commands:

Reviewer: ```suggestion
You can get the application's information with the following two commands:
```

```shell
# List all deployed applications
epinio app list

# Display the applications information
epinio app show mysimpleapp
```

### View installation logs

> If your application couldn't be installed, you might want to check your installation logs or, even better, save them into a file for a better screening with text editor.

Reviewer 2: > you might want to check your installation logs

"you might want to check your staging logs" might read better in context?

Reviewer: ```suggestion
If your application couldn't be deployed, you might want to check your staging logs or, even better, save them into a file for a better screening with text editor.
```

You can access the logs by running the command:

```shell
# Get the installation logs
epinio app logs --staging mysimpleapp

# Get the application logs

> epinio app logs --follow mysimpleapp

Reviewer 2: This is shown in more detail below, I'd pull this line out

```

### View application logs

> Another type of logs that you can have access is the application logs. And specially with web applications, you might want to have realtime logs displaying so you spot bugs faster.

Reviewer: ```suggestion
Another type of logs that you can access is the application logs. And specially with web applications, you might want to have realtime logs displaying so you spot bugs faster.
```
```suggestion
Another type of logs that you have access to is the application logs. And specially with web applications, you might want to have realtime logs displaying so you spot bugs faster.
```

> Epinio can display the logs both statically or dynamically as follow:

Reviewer: ```suggestion
Epinio can display the logs both statically or dynamically as follows:
```

```shell
# Display logs statically
epinio app logs mysimpleapp

# Display logs dynamically
epinio app logs --follow mysimpleapp
```

### Create a new port-forward

As described above, Epinio creates a new `ingress route` for your application. The route is bound by default to the port `443`.

However, you might need to test parts of your application using a different port. For these specific cases, you can run the following command:

```shell
epinio app port-forward 8080:8080 mysimpleapp
```

:::tip

> You can specify only one port number. In that case, Epinio will open the port of both `local` and `remote` targets.

Reviewer 2: It might be worth noting that by default all applications are listening inside of their pod on port 8080 so specifying both is useful for when you are forwarding multiple apps at the same time.

Author: I do agree, however as port forwarding will throw an error if the port is already used on the host, then I think it's a bit outside the scope of this tutorial.

For more information, you can see the [Port Fowarding page](../howtos/port_forwarding.md).

:::

### Scale your application

Another common task with Cloud Native applications, is to add (and remove) several instances of your application. This feature, called scaling, can be achieved with Epinio with the following command:

```shell
epinio app update mysimpleapp --instances 3
```

After you scaled your application up or down, you can check the status with the command:

```shell
epinio app show mysimpleapp
```

> ## Uninstall your application

Reviewer 2: I feel like "uninstall" is the wrong word here?

Reviewer: ```suggestion
## Remove your application
```

The whole process is handled by Epinio, which enables you to concentrate on your application rather than knowing how you'll be able to deploy it.

Let's see first how to deploy a simple application:

```shell
# Example code: https://github.com/epinio/example-12factor

# Move to the source code directory. Here is an example:
cd /github/example-12factor

# Deploy your application
epinio push -n mysimpleapp
```

At the end of the deployment output, you have the URL to be used for checking your application:

```shell
Deploying application ...
..............
🕞  Creating application resources
✔️  App is online.
Name: mysimpleapp
Namespace: workspace
Builder Image:
Routes:
1: https://mysimpleapp.127.0.0.1.sslip.io
```

### List the applications deployed

If you're working on many applications, it can be really useful to see when was the last time they were deployed and which URL you should use to check them.

You can get the application's information with the following two commands:

```shell
# List all deployed applications
epinio app list

# Display the applications information
epinio app show mysimpleapp
```

### View installation logs

> If your application couldn't be deployed, you might want to check your staging logs or, even better, save them into a file for a better screening with text editor.

Reviewer: ```suggestion
If your application couldn't be deployed, you might want to check your staging logs or, even better, save them into a file for a better screening with a text editor.
```
